### DISCUSSION TITLE
Adding wifi setup button

### DISCUSSION THREAD

**Discussion by leotordo** [HomeSpan/HomeSpan#466] [Q&A]
Hello everybody.
I'm new on Homekit.
I tried HomeSpan examples con a ESP32 board and it work.
I've a question about "Example 15: Real PushButtons".
I built a board with three buttons and all works well.
Is it possible to add another button for launch _HomeSpan’s temporary WiFi network_ and the related procedure?
Thanks
Stefano

**Comment by HomeSpan** [comment]
Yes, this functionality is already built into HomeSpan and works with any sketch, even those that don't use pushbuttons for HomeKit purposes.  See the discussion on the [Getting Started](https://github.com/HomeSpan/HomeSpan/blob/master/docs/GettingStarted.md) for details about adding a HomeSpan Status LED and HomeSpan Control Button, which will, among other functions, allow you to start the temporary WiFi network.

**Reply by leotordo** [reply]
Hello,
 I tried your suggestion. Ok. All works.
In my application I've a pic microcontroller for control other features.
I want to setup wifi and unpair device from microcontroller.
I can act con GPIO0 and emulate the 10s control button press for the Factory Reset, if needed.
If I want to run the HomeSpan’s temporary WiFi network from my microcontroller the only way I found is to emulate the correct Device Configuration Mode sequence (3s press than press twice, than long press). Is this the only method?
Thanks
Stefano

**Reply by ghost** [reply]
No, you should be able to simply run `processSerialCommand("A");` if I am not mistaken

**Reply by HomeSpan** [reply]
Yes, that's exactly right.  You can programmatically run any of the CLI commands this way - sort of an undocumented feature.

**Reply by leotordo** [reply]
thanks all. I tried on CLI. Perfect.
Ok for code  _processSerialCommand("A");_
Now....how can I implement the _if_ condition on a general GPIO?
Sorry but I'm very poor with this kind of program language
Thanks a lot
Stefano

**Reply by leotordo** [reply]
> Yes, that's exactly right. You can programmatically run any of the CLI commands this way - sort of an undocumented feature.

can I send "A" char from microcontroller UART directly to EPS32?
Thanks
Stefano

**Reply by ghost** [reply]
I am not sure about UART, but this is basic Arduino stuff and not really related to this library.
You could check it here: https://docs.arduino.cc/built-in-examples/digital/Button and just replace 
digitalWrite with processSerialCommand

**Reply by leotordo** [reply]
> I am not sure about UART, but this is basic Arduino stuff and not really related to this library. You could check it here: https://docs.arduino.cc/built-in-examples/digital/Button and just replace digitalWrite with processSerialCommand

thanks for the suggestion. I'll try it first because the UART test needs some track cut on my board...I'll test UART from MCU on next prototype
Regards
Stefano

**Comment by leotordo** [comment]
I added this instrutions

in setup
`pinMode(33, INPUT);`

in loop

`void loop(){`
  `  homeSpan.poll();`
`  if (digitalRead(33) == LOW) {`
  `    processSerialCommand("A");`
` }`
`} // end of loop()`

but compiler give me this error message

_C:\Users\...\LED_DIM2.ino: In function 'void loop()':_
_C:\Users\...\LED_DIM2.ino:164:5: error: 'processSerialCommand' was not declared in this scope_
     _processSerialCommand("A");_
     _^~~~~~~~~~~~~~~~~~~~_
_C:\Users\...\LED_DIM2.ino:164:5: note: suggested alternative: 'SpanUserCommand'_
     _processSerialCommand("A");_
     _^~~~~~~~~~~~~~~~~~~~_
     _SpanUserCommand_
_exit status 1_
_Compilation error: 'processSerialCommand' was not declared in this scope_

Did I write wrong code?
Thanks
Stefano

**Reply by leotordo** [reply]
> I added this instrutions
> 
> in setup `pinMode(33, INPUT);`
> 
> in loop
> 
> `void loop(){` ` homeSpan.poll();` ` if (digitalRead(33) == LOW) {` ` processSerialCommand("A");` ` }` `} // end of loop()`
> 
> but compiler give me this error message
> 
> _C:\Users...\LED_DIM2.ino: In function 'void loop()':_ _C:\Users...\LED_DIM2.ino:164:5: error: 'processSerialCommand' was not declared in this scope_ _processSerialCommand("A");_ _^~~~~~~~~~~~~~~~~~~~_ _C:\Users...\LED_DIM2.ino:164:5: note: suggested alternative: 'SpanUserCommand'_ _processSerialCommand("A");_ _^~~~~~~~~~~~~~~~~~~~_ _SpanUserCommand_ _exit status 1_ _Compilation error: 'processSerialCommand' was not declared in this scope_
> 
> Did I write wrong code? Thanks Stefano

ok solved

it's `homeSpan.processSerialCommand("A");`

waiting from HomeSpan Maintainer some infos on direct UART command by external MCU
bye
Stefano
